README
Decuda sm_1_1 (G80) disassembler version 0.4.x

Wladimir J. van der Laan <laanwj@gmail.com>, 2007-2009

This is my shot at providing an disassembler for the raw G8x cubin
instructions,  It was a big mystery for everyone, as NVIDIA is unwilling to
provide any information, but after a lot of experimentation it appears I 
figured out most of it.

Note that I acquired all the information by differential analysis on
the cubin files produced by ptxas and extensive experimentation on the
results. It is not based on reverse engineering of hardware or software.

I tried to mimic <i>PTX<i> instructions as closely as possible in the output, 
for an overview of the <i>PTX</i> assembly language see PTX_ISA_x.x.PDF in 
the doc directory of the NVIDIA CUDA package.

Note that this is an initial version, probably full of bugs. Let me know 
if you find any problems or have questions.

== Usage ==

Using the disassembler

Usage: decuda.py [OPTION]... [FILE]...

-p, --plain             Don't colorize output
-h, --help              Show this help screen
-k, --kernel-name <x>   Disassembly kernel name <x> of the cubin file
-n, --kernel-number <x> Disassemble kernel number <x> of the cubin file
-o, --output-file <x>   Output the kernel in assembler-compatible format to a file

Standard is ANSI colorized output, pipe through less -r to view this most
effectively. If color is not desired, use the -p flag.

Starting from version 0.3.0, decuda has an extensive internal debugging 
system and  logs each bit of the instruction that is visited (or not). 
Unvisited bits are assumed to be 0. If not, a warning is issued like 
'unk0 00400000'. This means that one of the high bits of the first 
instruction word is unaccounted for. Usually, this means the instruction 
has some slightly other meaning than decuda thinks it is: different 
operands, alternative interpretations, ... Please
let me know if you find out the meaning of unknown bits.

Using the assembler:

Usage: cudasm [OPTION]... [FILE]...

-h, --help              Show this help screen
-k, --kernel-name <x>   Name of the kernel to assemble
-o, --output-file <x>   Name of output file

Converts an assembler text file to a .cubin binary file that can be used
with the CUDA API.

Note: the assembler is in beta stage, although I've taken care to verify
that instructions assemble back into a form that is disassembled succesfully
by the disassembler, it is very possible that I'm missing some bits and
pieces or got the arguments wrong.

== TODO ==

- Things to check (assembler)
    - texture instructions can have unk1.00000004
    - there is still some mad madness left (overlapping bitfields?)
    - what does sign bit in setf do? I appears that it must always be set, if not,
      it returns 0 in every case (except 0,0x7FFFFF/0x800000 it appears)
    - BF_CONSTSEG_X/BF_CONSTSEG (does every instruction support 4 bit constseg?)
    - bar.sync has various parameters: at least the barrier id (0-15?), but
      also at least two bitfields, if which one is always 0xfff. This is
      interesting because it might allow for synchronisation of only a part of
      the block instead of the whole thing. Each bit could be 32 or 64
      threads.
    - op3_imm flag, where can it be used? only in shl/shr/set/...?
    - BF_ALT bit on full fadd instruction, what does it do?
    - what is halfop 0x70? looks like another mad
    - mad(integer) instruction, if both operands are flipped (0x0c000000),
    something special happens; add carry maybe?
- Assembler
    - Warn or error when operands get ignored: predicate in, operand
      flip/invert*
    - More specific error reporting in general
    - Texture instructions are not parsed and not assembled yet
    - Constant data is not parsed yet
- Fill in constants, if they refer to the local segment (c1). Make this
  optional.
- Our own CUDA-like compiler. http://llvm.org/ looks like a good open source
  replacement for nvcc and ptx. It handles the optimization and parsing,
  making actual code is a matter of writing a backend.
- Chip-level emulator?

== G80 Hardware ==

There are 32 and 64 bit instructions. Most instructions are 64 bit, but sometimes 
two simple instructions can be fit into one 64 bit word. (how this relates
to dual-issue, I don't know).

There are two classes of instruction: flow control instructions, and normal
instructions. Flow control instructions are bra, return, call, join, trap
and others. These all have an immediate (or no) argument and don't work on
instructions. Normal instructions do calculations and memory access.

G8x can address various different memory spaces:

- Registers. These are addressed with $r00 .. $r7f. Register $7f seems to be
a special 'bit bucket' register when no output is desired. Registers can be
addressed as 32 bit full register, or as 16 bit half (.lo and .hi).

- Shared memory, consisting of the grid parameters like blockIdx,
threadIdx and others. It also contains the program parameters, and shared
variables you declare. This is addressed using s[0x...].

Shared memory is prefixed by a header of 16 bytes, which contains
8 unsigned shorts. These are %gridflags, %ntid.*, %nctaid.x|y,
ctaid.x|y. Apart from these parameters being there it is normal, writable
shared memory.

- There are up to 16 constant memory segments, one for constants local to the kernel (segment 
1) and one for "global" constants (segment 0). Another (segment 14) is used
for relocations of global memory variables.
In contrary to common belief, these are not global in the program, but in the cubin file. It appears
they are limited to 64k. These are addressed using c0[0x...] and c1[0x...]
and so on.
This is analogous to the DirectX10 constant buffer. Here the 16 segments
are called slots.

- Then there is global and local memory that are accessed in a similar way,
which is to be expected as they are the same from the hardware side. These
are adressed using g[0x....] and l[0x....] respectively. Reads and writes to
this memory can be 8, 16, 32, 64 or 128 bits.

There are 16 global memory segments, just like there are 16 constant memory
segments, but they are independent. CUDA only uses segment 14 (0xe). Why 16 segments
are supported is unknown to me. Maybe it is used when loading from memory
buffers in DirectX 10 as well.

Addressing memory is quite interesting. It appears that there are four offset registers
(on current hardware) $ofs1, $ofs2, $ofs3 and $ofs4. These are used to add an
offset to the immediate offsets encoded in the instructions. Offsets can be loaded
by providing an immediate value, by adding an immediate value, or by using a register and 
a shift value. There are three bits reserved for the offset register, of
which 000 means use no offset. In theory there is place for 7 offset
registers. Ptxas only ever uses four though.

There are also a few internal values that can be read out: %physid, %clock,
%pm0, %pm1, %pm2 and %pm3. 

- %physid contains information about the current thread. In the topic 
  http://forums.nvidia.com/index.php?showtopic=45458 it is
  tried to find out what this one extactly means. It appears the format is 
  multiprocessor_ID*0x10000+warp_ID*0x100+position_in_warp
  where multiprocessor_ID seems to be one of the following 16 values:
  00 10 20 30 40 50 60 70 01 11 21 31 41 51 61 71
  and position_in_warp is 0x00 to 0x1F
- %clock is a cycle counter.
- %pm0..%pm3  (power mode?)

A special kind of register is the predicate register. G80 supports four of
these. Almost every instruction can output a result to a predicate register.
Instructions can be predicated so that they are only executed when that
predicate is true (@$p0), or false (!$p0). Apart from being true and false
it appears there are some more condition codes, but I haven't found out what
they mean yet.

The most complex instruction appears to be the tex instruction. The G80 can
sample from 1D, 2D and 3D textures using integer or floating point
coordinates. It can output one to four values, arbitrarily selected from 
{r,g,b,a}. The instruction exposed in CUDA is the basic non-mipmapped 
texture fetch. It would be interesting to know what the opcodes of the other
textureing instructions are, but unless we find a way to get the raw cubin
of a shader, we're stuck.

CUDA programs 'live' in a sandbox private memory space of 4GB. As the G80
has virtual addressing, memory regions that the CUDA context needs for
constants, global memory and shader code are mapped into this space. Cubin
code is DMA'ed to the card unchanged, with the exception of flow control
instructions which are relocated to their new address in the sandbox.

- global loads and stores take a whole-register argument

- set instruction: return 0xffffffff if condition holds, 0x00000000 otherwise

- flags (predicicate)
1  Z  zero flag
2  S  sign flag
4  C  carry flag
8  O  overflow flag

For a table of the possible flag combinations against condition codes, see
Logic.py. logigen.py in the root directory generates this table according to
the description in the G80 opengl shader reference, 
http://developer.download.nvidia.com/opengl/specs/g80specs.pdf . It appears
this matches, apart from condition ne.

Conditions on zf: zf is ne, nzf is equ

- assembler supports 'd.u32' to insert raw instruction data
- half (32 bit) instructions must always be issued in pairs
- half instructions: oper3 high bit is const segment (0 or 1)
- offset registers have increment bit

== Credits ==

asadafag -- testing kernels, input on unknown bit fields, and general suggestions.
Sylvain Collange -- information on mad instruction and help with test cases.
#nouveau -- help with intercepting the shader code for fragment/vertex/geometry programs

== Citings ==

Some interesting papers to read that use decuda for their work:

- Barra, a Modular Functional GPU Simulator for GPGPU
Sylvain Collange and David Defour and David Parello
http://hal.archives-ouvertes.fr/docs/00/37/47/15/PDF/stats_on_instruction.pdf

- Benchmarking GPUs to tune dense linear algebra
Vasily Volkov and James W. Demmel
Proceedings of the 2008 ACM/IEEE conference on Supercomputing
http://portal.acm.org/citation.cfm?id=1413402

Mail me if you know anything I forgot to include.

== Legal ==

See the LICENSE file.

Do whatever you like with this source, use it to speed up your CUDA kernels,
make debuggers, emulators or other interesting projects. But please let me
know if you do.


ChangeLog
2009-08-20  Sylvain Collange <sylvain.collange@univ-perp.fr>
	* Fixed number of bits of operand 4 constant segment
	* Added Abs/Sat and rounding mode modifiers to cvt0 (I2I)
	* Fixed number of bits of operand 3
	* Added mad24.s16 and mad24 with carry
	* Reduced madness in mad.f32/mac.f32 (hopefully)
	* Added .atom flag for atomic shared memory moves (SM 1.2)
	* Added double-precision instruction decoding (SM 1.3)

2006-06-20  Wladimir J. van der Laan <laanwj@gmail.com>

	* Fix command line options, especially -k
	* Default to plain not ansi output

2008-06-11  Wladimir J. van der Laan <laanwj@gmail.com>

	* Rename decuda to decuda.py and cudasm to cudasm.py,
	  this is more convenient for windows users.

2008-01-03  Wladimir J. van der Laan <laanwj@gmail.com>

	* Add 0x12 instruction (load from shared) to assembler and
	  disassembler

2007-12-22  Wladimir J. van der Laan <laanwj@gmail.com>

	* Fix deltaf instruction for disassembler

2007-12-07  Wladimir J. van der Laan <laanwj@gmail.com>

	* re-assembly fixes
	* support inversion (~) on logic ops
	* assembler operand lookup restructured
	* factor out important constants from assembler rule table, this
	makes it easier to just cache the data structure later on, for
	improved load time.
	* Release 0.4.1

2007-12-06  Wladimir J. van der Laan <laanwj@gmail.com>

	* add all half instructions into instruction table
	* fix flip bits for half add instruction
	* simplify operand specification in instruction table
	* finish rules for all halfops except tex
	* finish rules for all immops
	* add .half on disassembly of half instructions (for easier
	re-assembly)
	* remove debugging stuff from CubinFile, change Formatter in
	preparation to re-parse segments
	* Immediate values ending on 0xf were being rejected, solved
	* disassemble bar.sync in a sane way

2007-12-05  Wladimir J. van der Laan <laanwj@gmail.com>

	* found out and/xor/or/not instruction format
        * fix not instruction
	* add support for .half 32 bit instructions (for now, only add and
        mov)
	* fix constant segment in half instructions
	* post-increment offset registers ($ofs+=4)
        * remove [$ofsX+$rX] addressing mode as it doesn't exist

2007-12-03  Wladimir J. van der Laan <laanwj@gmail.com>

	* cvt.neg instruction added

2007-12-02  Wladimir J. van der Laan <laanwj@gmail.com>

	* "Immediate operand 3" is only allowed on a few instructions, mad
	  and others that use subsubop bits as an input are certainly not
	  one of them.
	* Version 0.4.0

2007-12-01  Wladimir J. van der Laan <laanwj@gmail.com>

	* Improve error reporting of assembler somewhat

2007-11-30  Wladimir J. van der Laan <laanwj@gmail.com>

	* mul24 instruction fixes (quite nasty)
        * basic flow control
        * labels are resolved
        * predication (in and out)
        * assembler is now functional
        * format predicates and labels in re-parsable format
        * Logic.py contains all condition code / flag combinations

2007-11-29  Wladimir J. van der Laan <laanwj@gmail.com>

	* cvt instruction fixes (half/full reg confusion)
	* more assembler work
	* Instruction.__repr__() actually changed operands. Oops.
        * added cudasm
        * cubin file loading and writing factored out to CubinFile class

2007-11-27  Wladimir J. van der Laan <laanwj@gmail.com>

	* Finally found out the meaning of subsubop in ld.const: type, like
	  in shared memory load
        * Release 0.3.1

2007-11-25  Wladimir J. van der Laan <laanwj@gmail.com>

	* Assembler rules for operand selection cleaned up
        * st.shared disassembler cleaned up, bit meanings now known
        * there is no st or ld, only mov, without changes of perspective

2007-11-23  Wladimir J. van der Laan <laanwj@gmail.com>

	* Beginnings of assembler
        * Get rid of multiplier in s[....] and c[....]
        * d.u32 instruction to inject raw data
        * Fix bug in disassemble offset registers, where $ofs4 was
        erronously displayed as $ofs1
        * More addressing fixes

2007-11-18  Wladimir J. van der Laan <laanwj@gmail.com>

	* Update readme on global memory segments
        * promote OP_SIGN_NONE operands if one of the operands is signed
        * Interpretation of alt bit for add.u32 instruction was wrong, it
        flips the result, not one of the operands
        * Register naming to decimal

2007-11-13  Wladimir J. van der Laan <laanwj@gmail.com>

        * Instructions can source from constant segment 2 or 3 as well
        * Fill in condition codes
        * Output registers added (as used in vertex shaders)
        * Float16 conversion
        * Release 0.3.0
        * All condition codes complete (afaik)

2007-11-12  Wladimir J. van der Laan <laanwj@gmail.com>

	* decudaraw util - decode raw cubin code from commandline, useful
	for disassembling shaders in mmio logs.

2007-11-09  Wladimir J. van der Laan <laanwj@gmail.com>

        * subsubop and subop are now a property instead of an attribute.
	This helps find cases in which they are missed.
	* Added an extensive internal debugging system to account for 
	each bit set in instruction words. This will, ultimately, help
	a lot when writing the assembler counterpart.
        * Heed 'ignore result' bit, no more register $r7f.
        * addc (add with carry) instruction.
        * add rounding op for add.f32.
        * sop.4 and sop.5 defined
	* some instructions are not predicated (nop, bar, some flow control)
        * rounding mode for mad

2007-11-08  Wladimir J. van der Laan <laanwj@gmail.com>

        * Last argument of mad.f32 comes from register
        * Release 0.2.9
        * remove self.flags, this served no purpose beyond to confuse
        bitfields
        * solve an issue with constant operands (hopefully)
        * self.bits(i,mask) instead of manually and/shift to clean up code
        * mark visited bits, warn if we neglected any

2007-11-03  Wladimir J. van der Laan <laanwj@gmail.com>


Disass.py
#!/usr/bin/python
# sm1_1 (G80) disassembler (decuda)
# Wladimir J. van der Laan, 2007
from cStringIO import StringIO
from Operand import *
from Instruction import *
from FlowControl import *
from Constants import *
from Opcodes import *
from Formatter import *
# Bits known 
# inst[0]:  
#   0xF0000000 op
#   0x0C000000 offset register bit 0 and 1
#   0x02000000 offset to immediate (0), offset to register (1)
#   0x01000000 oper3 from constant
#   0x00800000 oper2 from parameter
#   0x00400000 sign of oper3 (short instruction)
#   0x003F0000 oper3 (full instruction)
#   0x001F0000 oper3 (short instruction or immediate)
#   0x00008000 sign of oper1 (short instruction)
#   0x0000FE00 oper1
#   0x000001FC oper2
#   0x00000002 0=normal, 1=system (flow control)
#   0x00000001 0=32bit, 1=64 bit
# inst[1]:  
#   0xE0000000 subop
#   0x08000000 signed/unsigned
#   0x04000000 32/16
#   0x01000000 oper3 is immediate
#   0x03C00000 segment, on constant loading
#   0x00E00000 type, on load instructions
#   0x00800000 const segment bit 1
#   0x00400000 oper3/4 from constant in segment 1
#   0x00200000 oper2 from parameter
#   0x001FC000 oper4 or sub-sub op
#   0x00003000 predicate to act on
#   0x00000800 ??
#   0x00000400 ?? (usually set, unless predicated)
#   0x00000200 execute on pred
#   0x00000100 execute on !pred
#   0x00000080 ?? (usually set, unless predicated)
#   0x00000040 set predicate
#   0x00000030 predicate to set
#   0x0000003C atomic op (cannot set predicate)
#   0x00000008 also related tot set predicate
#   0x00000004 offset register bit 2
#   0x00000003 marker (0=normal,1=end,2???,3=immediate)

# Flow control ops (defined in Flowcontrol.py)
sys_ops = [
# 0,  1,    2,    3
Exit, Bra,  Bra,  Exit, 
# 4,  5,    6,    7
Bra,  Exit, None, None, 
# 8,  9,    A,    B
Sync, Exit, Bra,  None, 
# C,  D,    E,    F
None, None, None, None, 
]

# Subop matrix (defined in Opcodes.py)
ops = [
# 00      01      02      03      04      05      06      07
[None,   t01,    t02,   ldgpu,  None,    t05,    ldofs1, stsha],
# 10      11      12      13      14      15      16      17
[ld,     ldconst,ldshar,None,   None,   None,   None,   None],
# 20      21      22      23      24      25      26      27
[add,    None,   None,   None,   None,   None,   None,   None],
# 30      31      32      33      34      35      36      37
[neg,    None,   None,   m3x,    m3x,    m3x,    m3x,    m3x],
# 40      41      42      43      44      45      46      47
[mul24,  None,   None,   None,   None,   None,   None,   None],
# 50      51      52      53      54      55      56      57
[sad,    None,   None,   None,   None,   None,   None,   None],
# 60      61      62      63      64      65      66      67
[mad16,  mad16,  None,   mad24,  mad24,  mad24,  mad24,  mad24],
# 70      71      72      73      74      75      76      77
[None,   None,   None,   None,   None,   None,   None,   None],
# 80      81      82      83      84      85      86      87
[unk80,  None,   None,   None,   None,   None,   None,   None],
# 90      91      92      93      94      95      96      97
[flop,   None,   flop,   flop,   flop,   flop,   flop,   None],
# A0      A1      A2      A3      A4      A5      A6      A7
[cvt0,   cvt0,   cvt2,   satf,   cvt4,   None,   cvt6,   deltaf],
# B0      B1      B2      B3      B4      B5      B6      B7
[addf,   None,   None,   setf,   fmaxmin,fmaxmin,flop6,  None],
# C0      C1      C2      C3      C4      C5      C6      C7
[mulf,   None,   slct,   None,   None,   None,   None,   None],
# D0      D1      D2      D3      D4      D5      D6      D7
[logic,  ldofs0, ld_op,  st_op,  ld_op,  st_op,  atomic, atomic],
# E0      E1      E2      E3      E4      E5      E6      E7
[fmad,   None,   fmad,   addf,   mulf,   None,   None,   setf],	# ???
# F0      F1      F2      F3      F4      F5      F6      F7
[tex,    None,   None,   tex,    None,   None,   None,   nop],
]


class Disassembler(object):
    def decode(self, address, inst):
        """
        Find the type of an instruction, and call for it to be decoded.
        """
        i = Instruction()
        i.address = address
        i.inst = inst
        i.decode()
        
        # Pass on to specific op
        if i.system:
            decoder = sys_ops[i.op]
        else:
            decoder = ops[i.op][i.subop]
        if decoder == None:
            # unknown instruction
            # system, op, subop
            return i
        else:
            # Use specific decoder
            i = decoder(i)
            try:
                i.decode()
            except DecodeError, e: # Error decoding instruction
                i.warnings.append(e.message)
            return i

